### PR TITLE
feat: braket-harmony compilation

### PR DESCRIPTION
**Pull request by mmlouamri**
#233

## Changes:
I have added a new file in which I have implemented a function to compile a Braket circuit to IonQ `harmony`.

## Details or comments:
Here is an example of how the function works:
```python
# import the circuit module
from braket.circuits import Circuit

# define circuit with 4 qubits
my_circuit = Circuit().h(range(4)).cnot(control=0, target=2).cnot(control=1, target=3).ccnot(0, 1, 2)
print(my_circuit)
```
Gives:
```
T  : |0| 1 |2|
              
q0 : -H-C---C-
        |   | 
q1 : -H-|-C-C-
        | | | 
q2 : -H-X-|-X-
          |   
q3 : -H---X---

T  : |0| 1 |2|
```
The CCNOT is not supported by IonQ Harmony and thus it need to be compiled:
```python
compiled = braket_ionq_compilation(my_circuit)
print(compiled)
```
Gives:
```
T  : |0| 1 |2|3|4 |5|6|7|8 |9|10|11|12|
                                       
q0 : -H-C----------C--------C-C--T--C--
        |          |        | |     |  
q1 : -H-|-C---C----|---C-T--|-X--Ti-X--
        | |   |    |   |    |          
q2 : -H-X-|-H-X-Ti-X-T-X-Ti-X-T--H-----
          |                            
q3 : -H---X----------------------------

T  : |0| 1 |2|3|4 |5|6|7|8 |9|10|11|12|
```

## Checklist before merge
Add an `x` between the boxes braket that apply, meaning to complete the check before any merge happen, feel free to ask help if you are unsure about anything.

### General 
- [x] I have read the CONTRIBUTING doc.

- [x] ALL new features must include unit test to the test directory, codecov will check for any missing test code exist.

- [x] Ensure that all the test passes.

### CODE REVIEW

**Comment by mmlouamri** [comment]
Thank you for reviewing the PR and providing valuable feedback. I have incorporated the compilation function to the `AwsDeviceWrapper.run` method. Additionnaly, I have added a series of test to see if Braket's gate get correctly compiled (this tests the `braket_ionq_compilation` function). If necessary, I can also add tests to the `AwsDeviceWrapper` class to check that it is correctly integrated with the compilation function.

Please let me know if there are any further modifications or improvements you would like to see.

**Comment by ryanhill1** [comment]
@mmlouamri Awesome! I've just pushed some CI/CD update for unit tests. Could you pull / merge the latest changes from main and then I can review?

**Comment by mmlouamri** [comment]
> @mmlouamri Awesome! I've just pushed some CI/CD update for unit tests. Could you pull / merge the latest changes from main and then I can review?

I have merged the latest commits from main. I believe that the code is ready to be reviewed :).

File: qbraid/devices/aws/device.py [hunk 1]
```diff
@@ -232,6 +233,10 @@ def run(
 
         """
         run_input, qbraid_circuit = self._compat_run_input(run_input)
+
+        if self._provider_name == "IonQ" and self._name == "Harmony":
+            run_input = braket_ionq_compilation(run_input)
```
Comment (ryanhill1) [inline qbraid/devices/aws/device.py:238]: Here, the `braket_ionq_compilation()` conditional should actually be written into the `_compat_run_input()` method. The reason is that the compilation function already performs conversion to `pytket` and back to `braket`, so performing an extra conversion to `braket` beforehand is redundant.

**Review by ryanhill1 — CHANGES_REQUESTED** [review]
Thanks for the great PR! Your code looks great. I've ran a few tests on my end, and your method is working for the vast majority of cases. Although, I did find two `braket` gates for which it is failing:

```python
import numpy as np
from braket.aws import AwsDevice
from braket.circuits import Circuit

from qbraid.devices.aws.ionq import braket_ionq_compilation

circuit = Circuit()
circuit.phaseshift(0, np.pi / 8)
circuit.cphaseshift(0, 1, np.pi / 4)

device = AwsDevice("arn:aws:braket:us-east-1::device/qpu/ionq/Harmony")
compiled = braket_ionq_compilation(circuit)
task = device.run(compiled, shots=1)
task.cancel()
```

From the `phaseshift` gate, I get the following error:
```python
...
botocore.errorfactory.ValidationException: An error occurred (ValidationException) when calling the CreateQuantumTask operation: [line 4] uses a gate: phaseshift which is not supported by the device or defined via a defcal
```

And from the `cphaseshift` gate, I get the following error:
```python
...
botocore.errorfactory.ValidationException: An error occurred (ValidationException) when calling the CreateQuantumTask operation: [line 4] arbitrary unitary gates are not supported on the requested device
```

Take a look and let me know if you're getting the same thing. One possible source of the error(s) could be during our `pytket` $\rightarrow$ `braket` conversion at the end. If this is the case, one work around could be using the `pyket-braket` [`tk_to_braket`](https://github.com/CQCL/pytket-braket/blob/a6242d593f006587057413b8dcabd3a41705193d/pytket/extensions/braket/braket_convert.py#LL37C5-L37C17) function instead of our own usual circuit wrapper / transpiler.

File: qbraid/devices/aws/device.py (hunk 1 above)
Comment (mmlouamri) [reply qbraid/devices/aws/device.py]: Hello,

I have changed the transpiler from qBraid to pytket-braket and the above gates are getting compiled. In addition to that, I have put the conditional inside the `_compat_run_input`, now it doesn't transpile from pytket -> braket -> pytket -> braket anymore.

**Review by ryanhill1 — APPROVED** [review]

**Comment by ryanhill1** [comment]
@mmlouamri Merged, congrats! Can you please once again comment on the original issue (https://github.com/qBraid/qBraid/issues/233) so I can assign it to you and you can receive the credit for Unitary Hack?

**Pull request merged**